### PR TITLE
[Fix #595 #612] Fix infinite loop when reverse searching for next definition

### LINKED ISSUE #595: Syntax `^:` causes Emacs to freeze.
This happened to me about a half a year ago on Arch Linux with obviously older Emacs and older clojure-mode. Both using Prelude. I wasn't able to reproduce merely with clojure-mode, so steps include downloading prelude.

## Expected behavior
Adding ```^:``` to def and defn don't freeze emacs

## Actual behavior
```^:``` freezes emacs

## Steps to reproduce the problem
Download Prelude Emacs. Open either a clj or cljs file. Type ```(def ^:something)``` or ```(defn ^:something)``` or even just ```(def ^:)```/```(defn ^:)```. I've reliably been able to reproduce if it doesn't happen initially by typing ```(defn ^:something)``` and then M-backspace to delete back to the colon. Will include a screen shot of cpu profiler. 

Screen Shot 2021-05-16 at 2 20 43 PM

I know a screenshot is not necessarily the best, but I'm having trouble logging it otherwise.

### clojure-mode version

clojure-mode (version 5.13.0)

### Emacs version

GNU Emacs 28.0.50 (build 1, x86_64-apple-darwin19.6.0, NS appkit-1894.60 Version 10.15.7 (Build 19H114))
 of 2021-04-24

### Operating system

MacOSX 10.15.7 Catalina

### PR DESCRIPTION
**Pull request by OknoLombarda**
> Fixes https://github.com/clojure-emacs/clojure-mode/issues/595
> Fixes https://github.com/clojure-emacs/clojure-mode/issues/612

Fix infinite loop in `clojure-match-next-def` when typing metadata in def forms. Some more context [here](https://github.com/clojure-emacs/clojure-mode/issues/612#issuecomment-1186152534).

-----------------

Before submitting a PR mark the checkboxes for the items you've done (if you
think a checkbox does not apply, then leave it unchecked):

- [x] The commits are consistent with our [contribution guidelines][1].
- [x] You've added tests (if possible) to cover your change(s). Bugfix, indentation, and font-lock tests are extremely important!
- [x] You've run `M-x checkdoc` and fixed any warnings in the code you've written.
- [x] You've updated the changelog (if adding/changing user-visible functionality).
- [ ] You've updated the readme (if adding/changing user-visible functionality).

Thanks!

[1]: https://github.com/clojure-emacs/clojure-mode/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: clojure-mode.el [hunk 1]
```diff
@@ -754,8 +754,10 @@ Called by `imenu--generic-function'."
               (when (char-equal ?\) (char-after (point)))
                 (backward-sexp)))
           (cl-destructuring-bind (def-beg . def-end) (bounds-of-thing-at-point 'sexp)
-            (if (char-equal ?^ (char-after def-beg))
-                (progn (forward-sexp) (backward-sexp))
+            (when (char-equal ?^ (char-after def-beg))
+              (progn (forward-sexp) (backward-sexp)))
```
Comment (vemv) [inline clojure-mode.el:758]: While we're here, please add a comment explaining what this accomplishes

File: test/clojure-mode-syntax-test.el [hunk 2]
```diff
@@ -60,6 +60,36 @@
...
+    (with-clojure-buffer "
+(def a 1)
+(list [1 2 3])"
+      (end-of-buffer)
+      (clojure-match-next-def)
+      (expect (looking-at "(def"))))
+  (it "handles empty def forms"
+    (with-clojure-buffer "
+(def)
+(list [1 2 3])"
```
Comment (vemv) [inline test/clojure-mode-syntax-test.el:88]: Following https://github.com/clojure-emacs/clojure-mode/issues/612, there should be at least these extra cases:

* `(def ^{:macro})`
* `(def ^{:macro true})`
* `(def ^{:macro true} foo)`

And perhaps `(def ^{:macro} foo)` (which is both invalid and unlikely to type) too, especially if it doesn't complicate things.

File: test/clojure-mode-syntax-test.el [hunk 3]
```diff
@@ -60,6 +60,36 @@
       (expect (non-func "^hint " form) :to-be nil)
       (expect (non-func "#macro " form) :to-be nil))))
 
+(describe "clojure-match-next-def"
+  (it "handles vars with metadata"
+    (with-clojure-buffer "
+(def ^Integer a 1)
```
Comment (vemv) [inline test/clojure-mode-syntax-test.el:66]: Some other test cases that should be analog to this one:

* `(def ^:a a 1)`
* `(def ^::a a 1)`
* `(def ^::a/b a 1)`

**Review by vemv — COMMENTED** [review]
Looking reasonable!

Some suggestions for code coverage, will eye the impl again afterwards

File: test/clojure-mode-syntax-test.el (hunk 3 above)
Comment (OknoLombarda) [reply test/clojure-mode-syntax-test.el]: Done. I've also added `(def ^{:macro true} a 1)`

File: test/clojure-mode-syntax-test.el (hunk 2 above)
Comment (OknoLombarda) [reply test/clojure-mode-syntax-test.el]: Done

File: clojure-mode.el (hunk 1 above)
Comment (OknoLombarda) [reply clojure-mode.el:759]: Done

**Comment by vemv** [comment]
Great work!

I verified that if one discards the fix but keeps the tests, said tests indeed loop forever: https://app.circleci.com/pipelines/github/clojure-emacs/clojure-mode/261/workflows/9315c32d-6ac0-40e0-a7ef-e16eb51218d2/jobs/899

**Pull request merged**